src/main/java/com/fasterxml/jackson/core/util/TextBuffer.java

```diff
@@ -809,6 +810,7 @@ public String setCurrentAndReturn(int len) {
         }
         // more common case: single segment
         int currLen = _currentSize;
+        validateStringLength(currLen);
```

Comment:
My thinking here (and later points) is that we could actually calculate maximum allowable content; and if necessary, limit new segment size to allow that maximum plus one -- but given we allow "one too many", also consider possibility that caller may exceeded capacity (and we can throw exception now).
This would increase accuracy without too much extra work.

Reply:
I'm not sure. The text could be at the limit or slightly less. We end up asking for a new buffer extension when we are close to the limit - we won't hit the limit though - so I think it is best not to throw yet. I much happier with concept that the string limit means we can fail at the limit+1 or more but we never fail for limit, limit-1 or less.

Reply:
Sure, I did not mean to imply failing at anything not above limit, apologies for confusion.

What I simply meant was that when returning the next segment we can limit segment length to be such that when it fills it will be at minimum (or close) level.

Then again, considering maximum segment length is 64kchars I guess there's not much point in trying to optimize. At most we are late by 64k (128kB memory usage)